File: jdba-core-test-db/src/main/java/org/jetbrains/jdba/TestEnvironment.java
package org.jetbrains.jdba;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Map;
import java.util.Properties;



/**
 * @author Leonid Bushuev from JetBrains
 */
public abstract class TestEnvironment {

  public static final String CONNECTION_STRING_VAR = "test-db";
  public static final String CONNECTION_STRINGS_VAR_PREFIX = "test-db-";
  public static final String CONNECTION_STRINGS_FILE = "test-db.properties";

  private static final Properties mySystemProperties = System.getProperties();
  private static final Properties myLocalProperties = readLocalProperties();
  private static final Map<String,String> myEnv = System.getenv();



  @NotNull
  public static String obtainConnectionString() {
    String connectionString = getPossibleConnectionString(getVar(CONNECTION_STRING_VAR, null));
    if (connectionString == null) {
      System.err.print("Failed to obtain Test DB connection string\n");
      System.exit(-128);
    }
    return connectionString;
  }


  @Nullable
  protected static String getPossibleConnectionString(final @Nullable String cs) {
    if (cs != null) {
      String[] f = cs.split(":", 2);
      if (f.length == 1) {
        String anotherPropertyName = CONNECTION_STRINGS_VAR_PREFIX + f[0];
        String anotherChance = getVar(anotherPropertyName, null);
        return getPossibleConnectionString(anotherChance);
      }
      else {
        return cs;
      }
    }
    else {
      // TODO try to scan possible strings
      return null;
    }
  }


  @Contract("_,null->_; _,!null->!null")
  public static String getVar(@NotNull final String name, @Nullable final String defaultValue) {
    String value = mySystemProperties.getProperty(name);
    if (value == null) value = myLocalProperties.getProperty(name);
    if (value == null) value = myEnv.get(name);
    if (value == null) value = defaultValue;
    return value;
  }


  private static Properties readLocalProperties() {
    Properties p = new Properties();
    File dir = getCurrentDirectory();
    while (dir != null && dir.isDirectory()) {
      File file = new File(dir, CONNECTION_STRINGS_FILE);
      if (file.exists()) {
        readLocalPropertiesFromFile(p, file);
        break;
      }
      else {
        dir = dir.getParentFile();
      }
    }
    return p;
  }

  @NotNull
  private static File getCurrentDirectory() {
    File dir = new File(".");
    try {
      dir = dir.getCanonicalFile();
    }
    catch (IOException ioe) {
      ioe.printStackTrace();
    }
    return dir;
  }

  private static void readLocalPropertiesFromFile(final Properties p, final File localPropertiesFile) {
    try {
      InputStream is = new FileInputStream(localPropertiesFile);
      try {
        p.load(is);
      }
      finally {
        is.close();
      }
    }
    catch (IOException io) {
      System.err.println("Failed to read file " + CONNECTION_STRINGS_FILE + ": " + io.getMessage());
    }
  }


}

File: dekaf-test-db/src/org/jetbrains/dekaf/TestEnvironment.kt
package org.jetbrains.dekaf

import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.util.*


/**
 * @author Leonid Bushuev
 */
object TestEnvironment {

    private val CONNECTION_STRING_VAR = "test-db"
    private val CONNECTION_STRINGS_VAR_PREFIX = "test-db-"
    private val CONNECTION_STRINGS_FILE = "test-db.properties"

    private val mySystemProperties = System.getProperties()
    private val myLocalProperties = readLocalProperties()
    private val myEnv = System.getenv()

    @JvmStatic
    fun obtainConnectionString(): String {
        val connectionString = getPossibleConnectionString(getVar(CONNECTION_STRING_VAR))
        if (connectionString == null) {
            System.err.println(
                    """|Failed to obtain Test DB connection string.
                       |The connection string is searched in the following locations:
                       |	1) '$CONNECTION_STRING_VAR' property in ./$CONNECTION_STRINGS_FILE;
                       |	2) '$CONNECTION_STRING_VAR' environment variable;
                       |	3) '$CONNECTION_STRING_VAR' system property.
                    """.trimMargin())
            System.exit(-128)
            throw RuntimeException() // just for compiler
        }
        return connectionString
    }


    @JvmStatic
    private fun getPossibleConnectionString(cs: String?): String? {
        if (cs != null) {
            val f = cs.split(":".toRegex(), 2).toTypedArray()
            if (f.size == 1) {
                val anotherPropertyName = CONNECTION_STRINGS_VAR_PREFIX + f[0]
                val anotherChance = getVar(anotherPropertyName)
                return getPossibleConnectionString(anotherChance)
            }
            else {
                return cs
            }
        }
        else {
            // TODO try to scan possible strings
            return null
        }
    }


    @JvmStatic
    fun getVar(name: String): String? {
        var value: String? = mySystemProperties.getProperty(name)
        if (value == null) value = myLocalProperties.getProperty(name)
        if (value == null) value = myEnv[name]
        return value
    }

    @Deprecated("Use the '?:' operator instead", ReplaceWith("getVar(name) ?: defaultValue", "org.jetbrains.dekaf.TestEnvironment.getVar"))
    @JvmStatic
    fun getVar(name: String, defaultValue: String): String = getVar(name) ?: defaultValue


    @JvmStatic
    private fun readLocalProperties(): Properties {
        val p = Properties()
        var dir: File? = currentDirectory
        while (dir != null && dir.isDirectory) {
            val file = File(dir, CONNECTION_STRINGS_FILE)
            if (file.exists() && !file.isDirectory) {
                readLocalPropertiesFromFile(p, file)
                break
            }
            else {
                dir = dir.parentFile
            }
        }
        return p
    }

    @JvmStatic
    private val currentDirectory: File
        get() {
            var dir = File(".")
            try {
                dir = dir.canonicalFile
            }
            catch (ioe: IOException) {
                ioe.printStackTrace()
            }

            return dir
        }

    @JvmStatic
    private fun readLocalPropertiesFromFile(p: Properties, localPropertiesFile: File) {
        try {
            val `is` = FileInputStream(localPropertiesFile)
            try {
                p.load(`is`)
            }
            finally {
                `is`.close()
            }
        }
        catch (io: IOException) {
            System.err.println("Failed to read file " + localPropertiesFile.path + ": " + io.message)
        }

    }


}
